### DISCUSSION TITLE
Best Practices for Organizing Code in Axum Workspace Setup

### DISCUSSION THREAD

**Discussion by NCura** [leptos-rs/leptos#2923] [Q&A]
Hi everyone,

I'm working with the Axum Workspace setup, which creates three members: `app`, `frontend`, and `server`. I have a question about how to best organize my code across these crates.

**What is the intended purpose of each crate (`app`, `frontend`, `server`) in this setup?** Specifically, what types of code should go into each?

For instance, if the `server` crate is meant for server-only code, what kinds of code or functionality should be placed there? Are there examples of server-only code that would fit well in the `server` crate? Initially, I thought that server functions should reside in the `server` crate. However, it was mentioned in another discussion that the `app` crate is the typical place to put server functions because components need to call these functions, and `server` depends on `app`, but not the other way around.

Currently, I'm placing everything inside the `app` crate because it simplifies things, but I realize this might negate the benefits of using a workspace setup. I’m interested in continuing with the workspace setup, even if it adds some complexity, as it forces me to confront issues that might be overlooked in a simpler setup.

I would really appreciate any guidance or examples on how to properly structure code across these crates.

Thank you!

**Comment by gbj** [comment]
The main benefit to using a workspace setup in this situation is that you do not need to recompile the `front` crate (your WASM frontend) when all you've done is made changes to unrelated parts of the server. For example, if you have API routes or other server logic that are not server functions, and which therefore don't require compiling the frontend crate. It can also be convenient for people who just prefer this code organization.

If you want to add a fourth crate for server-only functionality that will be imported by server functions in `app`, that is also completely possible.

It's purely a matter of individual preference.
